clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerHeartbeatRequestManager.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.internals.events.BackgroundEventHandler;
+import org.apache.kafka.clients.consumer.internals.metrics.HeartbeatMetricsManager;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.ConsumerGroupHeartbeatRequest;
+import org.apache.kafka.common.requests.ConsumerGroupHeartbeatResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+/**
+ * This is the heartbeat request manager for consumer groups.
+ *
+ * <p>See {@link AbstractHeartbeatRequestManager} for more details.</p>
+ */
+public class ConsumerHeartbeatRequestManager extends AbstractHeartbeatRequestManager<ConsumerGroupHeartbeatResponse> {
+
+    /**
+     * Membership manager for consumer groups
+     */
+    private final ConsumerMembershipManager membershipManager;
```

Comment:
maybe we can move `membershipManager` to be a variable of parent class (`AbstractHeartbeatRequestManager`)? we have the great abstract class `AbstractMembershipManager`

Reply:
I did try that but I need `ConsumerHeartbeatRequestManager.membershipManager()` to return an instance of `ConsumerMembershipManager` and not `AbstractMembershipManager<ConsumerGroupHeartbeatResponse>`. That is because the `ApplicationEventProcessor` calls methods of the consumer membership manager which are not in the abstract membership manager.

I think I could use a second type variable to AbstractHeartbeatRequestManager so it could return a strongly typed membership manager, but I didn't like that so much and settled on putting the membership manager references into the subclasses instead.